cli/derive/src/lib.rs

```diff
@@ -0,0 +1,173 @@
+            }
+        };
+
+        let expanded = quote! {
+            #[inline]
+            #[allow(clippy::redundant_pub_crate)]
+            pub(crate) fn #fun_name < O, E, F, Fut, Fil, #( #arg_types ),* > (
+                f: F,
+                router: Fil,
+            ) -> impl Filter<Extract = (WarpResult<O, E>,), Error = Rejection> + Clone
+            where
+                Fil: Filter<Extract = ( #( #arg_types ),* ), Error = Rejection> + Clone,
+                F: Fn( #( #arg_types ),* ) -> Fut + Copy + Send + Sync + 'static,
+                Fut: std::future::Future<Output = Result<O, E>> + Send,
+                #( #arg_types: Send ),*
+                {
+                    router.and_then(move | #( #arg_names ),* | async move {
+                        Ok::<_, Infallible>(WarpResult(f( #( #arg_names ),* ).await))
+                    })
+                }
+        };
+
+        endpoints.push(expanded);
+    }
+
+    quote! {
+        #( #endpoints )*
+    }
+    .into()
+}
+
+#[derive(Debug)]
+struct EndpointList(Vec<EndpointItem>);
+
+#[derive(Debug)]
+enum EndpointItem {
+    NameAndArgCount { name: LitStr, arg_count: LitInt },
+    ArgCount(LitInt),
+}
+
+impl Parse for EndpointList {
+    fn parse(input: ParseStream) -> SynResult<Self> {
+        let items = Punctuated::<EndpointItem, Token![,]>::parse_terminated(input)?;
+        let mut seen_arg_counts = Vec::new();
+        for item in items.iter() {
+            match item {
+                EndpointItem::NameAndArgCount { name, arg_count } => {
+                    let curr_count = arg_count.base10_parse::<u8>()?;
+                    if seen_arg_counts
+                        .iter()
+                        .any(|&old_count| old_count == curr_count)
+                    {
+                        return Err(syn::Error::new_spanned(
+                            arg_count.token(),
+                            "argument counts for all endpoints should be distinct",
+                        ));
+                    } else if name.value().contains(' ') {
+                        return Err(syn::Error::new_spanned(
+                            name.token(),
+                            "function name should not contain spaces",
+                        ));
+                    }
+                    seen_arg_counts.push(curr_count);
+                }
+                EndpointItem::ArgCount(arg_count) => {
+                    let curr_count = arg_count.base10_parse::<u8>()?;
+                    if seen_arg_counts
+                        .iter()
+                        .any(|&old_count| old_count == curr_count)
+                    {
+                        return Err(syn::Error::new_spanned(
+                            arg_count.token(),
+                            "argument counts for all endpoints should be distinct",
+                        ));
+                    }
+                    seen_arg_counts.push(curr_count);
```

Comment:
This is also duplicated in both match arms. Consider rewriting like this:

```rust
let count = match item {
    EndpointItem::NameAndArgCount { name, arg_count } => {
        // ...
        curr_count
    }
    EndpointItem::ArgCount(arg_count) => {
        // ...
        curr_count
    }
}?;
seen_arg_counts.push(count);
```
I think, `seen_arg_counts` modification is more clear here and match arms are doing only one thing -- parsing count

Reply:
Same as for duplication above